pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -343,18 +343,29 @@ func (t *Topology) countDomains(ctx context.Context, tg *TopologyGroup) error {
 	// Note: long term we should handle this when constructing the domain groups, but that would require domain groups
 	// to handle affinity in addition to taints / tolerations.
 	for _, n := range t.stateNodes {
-		// ignore state nodes which are tracking in-flight NodeClaims
-		if n.Node == nil {
-			continue
+		labels := n.Labels()
+		requirements := scheduling.NewLabelRequirements(labels)
+
+		domain, exists := labels[tg.Key]
+
+		// We special-case kubernetes.io/hostname primarily for stateNodes having NodeClaims without a node
+		//  and since their domain won't be registered until the node joins the cluster.
+		if !exists && tg.Key == corev1.LabelHostname {
```

Comment:
I'm confused why we need to fake the value in this code here -- why don't we just fake the value at the StateNode and then we don't have to special-case code in this scheduling section. I think all this fix amounts to is passing the same value that n.HostName() has back through n.Labels(). Seems like that would simplify a lot of this and you wouldn't have to make _any_ changes to this section of code